### DISCUSSION TITLE
Debugging using bpf_printk / bpf_trace_printk?

### DISCUSSION THREAD

**Discussion by XieGuochao** [cilium/ebpf#1083] [Q&A] [answered]
Hello, I want to do some simple observations first before developing tools. 

However, it seems that the `bpf_printk` or `bpf_trace_printk` does not work and nothing is printed on the terminal. Is it common? Is there any simple alternative to print something in the BPF program? Is there any guide for debugging/observation?

**Comment by lmb** [comment, marked as answer]
`bpf_printk` doesn't write to the console, it writes to the kernel's trace pipe. Take a look at https://man7.org/linux/man-pages/man7/bpf-helpers.7.html for entry `bpf_trace_printk`.

**Comment by liel-almog** [comment]
I've seen the @lizrice talk where she used `bpf_printl` and it printed to the screen using `libbpfgo` and the `bpf.TracePrint()`.
Do have the same functionality?

**Reply by ti-mo** [reply]
@liel-almog Assuming you mean `bpf_printk`, and given it doesn't look like `bpf.TracePrint()` is part of libbpfgo anymore, a `sudo cat /sys/kernel/debug/tracing/trace_pipe` will get you to the same place.

**Reply by lizrice** [reply]
If you have `bpftool` installed you can also use `bpftool prog tracelog` to show the same trace output. (It's doing the same thing, I just find it easier to remember than the location of the `trace_pipe` pseudofile!)

**Reply by liel-almog** [reply]
Thank you so much!
I do have it installed.
By the way @lizrice , I'm a big fan of your talk and work!
Helped me so much.
